File: BFS/Employee_Importance.cpp
/*
// Definition for Employee.
class Employee {
public:
    int id;
    int importance;
    vector<int> subordinates;
};
*/

class Solution {
public:
    int getImportance(vector<Employee*> employees, int id) {
        
        int len = employees.size();
        if(len == 0)
            return 0;
        int val = 0;
        queue<int> q;
        for(auto employee : employees)
        {
            if(employee->id == id)
            {
                val += employee->importance;
                for(auto subordinate : employee->subordinates)
                   if(subordinate)
                        q.push(subordinate);
            }
            
        }
        
        while(!q.empty())
        {
             for(auto employee : employees)
            {
                int id = q.front();
                if(employee->id == id)
                {
                    q.pop();
                    val += employee->importance;
                    for(auto subordinate : employee->subordinates)
                        if(subordinate)
                            q.push(subordinate);
                }

            }
        }
        return val;
        
    }
};




/* DFS */

/*
// Definition for Employee.
class Employee {
public:
    int id;
    int importance;
    vector<int> subordinates;
};
*/

class Solution {
    
    void DFS(unordered_map<int,Employee*> &map, int id,int &val)
    {
           auto employee = map[id];
            val += employee->importance;
             id = employee->id;
            for(auto sub : employee->subordinates)
            {
                if(sub)
                    DFS(map,sub,val);
            }
        
    }
public:
    int getImportance(vector<Employee*> employees, int id) {
        
        int len = employees.size();
        if(len == 0)
            return 0;
        int val = 0;
        unordered_map<int,Employee*> map;
        for(auto employee : employees)
            map[employee->id] = employee;
        DFS(map,id,val);
        return val;
        
    }
};

File: BFS/Java/Easy/Employee_Importance.java
/*
// Definition for Employee.
class Employee {
    public int id;
    public int importance;
    public List<Integer> subordinates;
};
*/

class Solution {
    public int getImportance(List<Employee> employees, int id) {
        
        int len = employees.size();
        Map<Integer,Employee> map = new HashMap<>();
        for(Employee e: employees)
            map.put(e.id,e);
        
        int totalImp =0;
        Queue<Employee> queue = new LinkedList<Employee>();
        queue.offer(map.get(id));
        while(!queue.isEmpty())
        {
            int size = queue.size();
            Employee currNode = queue.poll();
            totalImp += currNode.importance;
            for(int i : currNode.subordinates)
            {
                queue.add(map.get(i));   
            }
        }
        return totalImp;
        
    }
}